### PR TITLE
Prototype: BL support for .dfu file update from SD card 

### PR DESCRIPTION
**Pull request by hedger**
# What's new

- Added SD card & flash ops support to BL
- Added simple DFU file parser
- Validation of DFU file - supports only proper Flipper Zero firmware (no BL) images
- Able to erase, write & verify FW
- Toggle update: reboot to bootloader (back + left), then quickly press "down" button

# Verification 

- Build with SES (Makefile not updated for this draft)
- Put .dfu file to sd card "firmware/fw.dfu"
- Reboot to fw upgrade mode & verify that update process completes

# Checklist (do not modify)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: bootloader/src/dfu_file.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    if(!fs_file_is_valid(dfuf)) {
+        return 0;
+    }
+
+    f_rewind(&dfuf->file);
+
+    const uint32_t dfu_suffix_offset = dfuf->stat.fsize - sizeof(DfuSuffix);
+
+    CHECK_FRESULT(f_read(&dfuf->file, &dfu_prefix, sizeof(DfuPrefix), &bytes_read));
+    if((dfu_prefix.szSignature[0] != 'D') || (dfu_prefix.szSignature[1] != 'f') ||
```
Comment (skotopes) [inline bootloader/src/dfu_file.c:55]: More complex version of memcmp

File: bootloader/src/sd_update.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    .task_cb = &page_task_validate_flash,
+};
+
+static const DfuValidationParams flipper_dfu_params = {
+    .device = 0xFFFF,
+    .product = 0xDF11,
+    .vendor = 0x0483,
+};
+
+void sdcard_update_execute() {
```
Comment (skotopes) [inline bootloader/src/sd_update.c:87]: file - methods naming convention is a little bit off

File: bootloader/src/sd_update_progress.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#include "sd_update_progress.h"
+
+#include <furi_hal.h>
+
+static const char* update_state_text[] = {
```
Comment (skotopes) [inline bootloader/src/sd_update_progress.c:5]: Despite of the fact that we use this construction in some other places, I prefer having separatemethod that returns string by enum and have switch inside of it.


In the worst case use something like this:

```

static const Icon* ArchiveItemIcons[] = {
    [ArchiveFileTypeIButton] = &I_ibutt_10px,
    [ArchiveFileTypeNFC] = &I_Nfc_10px,
    [ArchiveFileTypeSubGhz] = &I_sub1_10px,
    [ArchiveFileTypeLFRFID] = &I_125_10px,
    [ArchiveFileTypeIrda] = &I_ir_10px,
    [ArchiveFileTypeFolder] = &I_dir_10px,
    [ArchiveFileTypeUnknown] = &I_unknown_10px,
};
```

File: bootloader/targets/f7/furi_hal/furi_hal_crc.c [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    .Init.CRCLength = CRC_POLYLENGTH_32B,
+    .Init.DefaultPolynomialUse = DEFAULT_POLYNOMIAL_ENABLE,
+    .Init.DefaultInitValueUse = DEFAULT_INIT_VALUE_ENABLE,
+    .Init.InputDataInversionMode = CRC_INPUTDATA_INVERSION_BYTE,
+    .Init.OutputDataInversionMode = CRC_OUTPUTDATA_INVERSION_ENABLE,
+    .InputDataFormat = CRC_INPUTDATA_FORMAT_BYTES,
+};
+
+void furi_hal_crc_init() {
+    __HAL_RCC_CRC_CLK_ENABLE();
```
Comment (skotopes) [inline bootloader/targets/f7/furi_hal/furi_hal_crc.c:16]: don't use full HAL, use LL

File: bootloader/src/sd_update_progress.c (hunk 3 above)
Comment (DrZlo13) [reply bootloader/src/sd_update_progress.c:5]: In my pet project, I have found that the following construction is most suitable for such cases.

```
static const char* error_texts[] = {
    [FR_OK] = "OK",
    [FR_DISK_ERR] = "DISK ERR",
    [FR_INT_ERR] = "INT ERR",
    [FR_NOT_READY] = "NOT READY",
    [FR_NO_FILE] = "NO FILE",
    [FR_NO_PATH] = "NO PATH",
    [FR_INVALID_NAME] = "INVALID NAME",
    [FR_DENIED] = "DENIED",
    [FR_EXIST] = "EXIST",
    [FR_INVALID_OBJECT] = "INVALID OBJECT",
    [FR_WRITE_PROTECTED] = "WRITE PROTECTED",
    [FR_INVALID_DRIVE] = "INVALID DRIVE",
    [FR_NOT_ENABLED] = "NOT ENABLED",
    [FR_NO_FILESYSTEM] = "NO FILESYSTEM",
    [FR_MKFS_ABORTED] = "MKFS ABORTED",
    [FR_TIMEOUT] = "TIMEOUT",
    [FR_LOCKED] = "LOCKED",
    [FR_NOT_ENOUGH_CORE] = "NOT ENOUGH CORE",
    [FR_TOO_MANY_OPEN_FILES] = "TOO MANY OPEN FILES",
    [FR_INVALID_PARAMETER] = "INVALID PARAMETER",
};

const char* fs_error_text(uint32_t error) {
    if(error >= COUNT_OF(error_texts)) {
        return "UNKNOWN";
    } else {
        return error_texts[error];
    }
}
```

**Pull request closed without merging**